doc/Neighbor/Neighbor_Refresh_HLD.md

```diff
@@ -0,0 +1,321 @@
+| **Term** | **Meaning**                                       |
+| -------- | ------------------------------------------------- |
+| ARP      | Address Resolution Protocol                       |
+| ND       | Neighbor Discovery                                |
+| IP       | Internet Protocol                                 |
+| L3       | Layer 3                                           |
+| RIF      | Routing Interface                                 |
+| VRF      | Virtual Routing and Forwarding                    |
+| SVI      | Switch Virtual Interface                          |
+
+# About This Manual
+This document describes the high level description of neighbor refresh in SONiC.
+
+# 1 Requirement Overview
+
+Refresh ARP/ND before their respective timer or MAC ageout.
+
+SONiC depends upon the Linux kernel to manage the ARP/ND tables. SONiC then listens to ARP/ND events from the kernel and synchronizes the hardware as required. Some of the limitations with this are as follows,
+
+- The kernel does not "see" the routed (in HW) through-traffic, and so cannot update its "hit bits" accordingly. Therefore the kernel may age out an entry that is still in use.
+- The kernel does not "see" the HW MAC aging process, and so does not know that a MAC address associated with an ARP/ND entry has been aged out, and so does not refresh it. This can result in traffic black holes for a "quiet" neighbor (i.e. one that does not transmit much) / unidirectional traffic streams.
+
+ARP Update helper process is used by SONIC to periodically refresh the kernel ARP table by sending ARP/ND packets to all neighbor table entries. This provokes responses from the associated hosts, and this keeps both the HW MAC table and kernel ARP tables current. However the current ARP Refresh process is implemented as a bash script and is not effective at scale. So, the proposal here is to design and implement a new ARP/Neighbor Refresh process.
+
+Earlier code PR https://github.com/Azure/sonic-swss/pull/1548 handled the neighbor refresh in neighsyncd. As per the review comments creating new HLD and Code PR for handling neighbor refresh in Nbrmgrd.
+
+## 1.1 Functional Requirements
+
+  - Refresh ARP/ND before the ARP/ND kernel ageout
+  - Refresh ARP/ND before the MAC ageout on Vlan interfaces
+  - On MAC delete (clear mac / control protocol) trigger refresh for corresponding ARP/ND
+
+## 1.2 Configuration and Management Requirements
+
+No New CLIs are added as part of this proposal.
+
+## 1.3 Scalability Requirements
+
+  - Support ARP/ND refresh for upto 4K Vlan Interfaces and 32K neighbor entries
+
+## 1.4 Warm Boot Requirements
+
+  - During Warmboot, MAC aging will be disabled and will be enabled back after the system is restared. Hence traffic blackholing due to MAC ageout will not happen. When the system boots up, ARP/ND refresh cache table will be populated again and then the neighbor entries will be refreshed as normal.
+
+## 1.5 Limitiations
+
```

Comment:
What happens when the ARP/Nbr is flushed in the kernel by the user? if any ARP/Nbr associated with the route, do we leave those routes to drop the packets in the HW or refresh the those nbrs alone?

Reply:
Orchagent will not delete the neighbor if there is a route associated, hence the HW will continue to forward the packet until the route is removed and then the neighbor will be removed from HW.